### PR TITLE
[bugfix] Resolve bug with multiple optimizers and toggle.

### LINKED ISSUE #5292: For multiple optimizers, model.toggle_optimizer() is setting the requires_grad=True for all the params.
https://github.com/PyTorchLightning/pytorch-lightning/blob/dabfeca92e0702e55f09ac53e9412672cd258cd3/pytorch_lightning/core/lightning.py#L1152-L1170

Setup : 
pytorch-lightning 1.1.2
pytorch 1.7.1

When using multiple optimizers, in the `toggle_optimizers(..)` function, the requires_grad property is set to true for all the params belonging to the param_groups of the optimizer. This is incorrect as in case the user has explicitly disabled the `requires_grad` property for some parameters permanently, the function would enable `requires_grad` for those parameters also. 

Proposed fix:
The requires_grad value for all the parameters should be stored beforehand in an object variable `self.params_dict` of lightning object. In the `toggle_optimizer()` function, instead of setting `params.requires_grad=True`, set `params.requires_grad = self.params_dict[params]`.This would set the correct value for the params of the concerned optimizer.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

In case of multiple optimizers, toggle optimizer was called.
It would lose user requieres_grad state by setting everything to True or False.

This is erroneous in case there are any parameters which are explicitly kept non-trainable.

Fixes #5292 #5642

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -1166,16 +1166,44 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
 
         Override for your own behavior
 
+        It works with `untoggle_optimizer` to make sure param_requires_grad_state is properly reset.
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:1169]:
```suggestion
        It works with ``untoggle_optimizer`` to make sure param_requires_grad_state is properly reset.
```

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -1166,16 +1166,44 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:1166]: maybe just use the opt_idx to get the optimizer down there... just to unify the arguments with `untoggle_optimizer`. Either way is fine.

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -1166,16 +1166,44 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
...
+                for group in opt.param_groups:
+                    for param in group['params']:
+                        param_requires_grad_state[param] = param.requires_grad
+                        param.requires_grad = False
 
         for group in optimizer.param_groups:
             for param in group['params']:
-                param.requires_grad = True
+                if param in param_requires_grad_state:
+                    param.requires_grad = param_requires_grad_state[param]
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1186]: Clean-up :broom: 

```suggestion
        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
            for group in opt.param_groups:
                for param in group['params']:
                    if optimizer_idx == opt_idx:
                        param.requires_grad = param_requires_grad_state.get(param, True)
                    else:
                        param.requires_grad = False
                        param_requires_grad_state[param] = param.requires_grad
```

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1166,16 +1166,44 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
...
+
+        self._param_requires_grad_state = param_requires_grad_state
+
+    def untoggle_optimizer(self, optimizer_idx: int):
+        """
+        .. note:: Only called when using multiple optimizers
+
+        Override for your own behavior
+
+        Args:
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1196]: Write the args or remove them. Same for `toggle_optimizer`

File: tests/core/test_lightning_module.py [hunk 5]
```diff
@@ -139,3 +140,77 @@ def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx, clos
...
+            self.layer_1[4].weight.requires_grad = False
+
+            self.layer_2[1].weight.requires_grad = False
+            self.layer_2[3].weight.requires_grad = False
+
+        def training_step(self, batch, batch_idx, optimizer_idx):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            return {"loss": loss}
+
```
Comment (carmocca) [inline tests/core/test_lightning_module.py:179]:
```suggestion
```

same as parent's

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -1166,16 +1166,44 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
...
+            optimizer_idx:
+        """
+        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
+            if optimizer_idx != opt_idx:
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1200]: shouldn't untoggle be done for all optimizers?

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1166]: Sounds great ! I will do so.

File: pytorch_lightning/core/lightning.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1186]: This won't work, but we will by sorting optimizer. Good suggestion ! I will do so !

Comment (jain-anshul) [reply pytorch_lightning/core/lightning.py:1186]: @carmocca  There seems to be an issue in the suggested edit for the line
`param.requires_grad = param_requires_grad_state.get(param, True)` 
This would set the `requires_grad` True for all the params of the current optimizer. That's exactly what we want to avoid here.

File: pytorch_lightning/core/lightning.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1200]: No, as toggle won't touch the current one.

File: pytorch_lightning/core/lightning.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1186]: Done !

File: tests/core/test_lightning_module.py (hunk 5 above)
Comment (tchaton) [reply tests/core/test_lightning_module.py:179]: Done !

Comment (carmocca) [reply tests/core/test_lightning_module.py:179]: Doesn't look done :)

File: tests/core/test_lightning_module.py [hunk 7]
```diff
@@ -139,3 +140,77 @@ def optimizer_step(self, current_epoch, batch_nb, optimizer, optimizer_idx, clos
...
+            if optimizer_idx == 1:
+                assert self.layer_1[0].weight.requires_grad is False
+                assert self.layer_1[2].weight.requires_grad is False
+                assert self.layer_1[4].weight.requires_grad is False
+
+                assert self.layer_2[1].weight.requires_grad is False
+                assert self.layer_2[3].weight.requires_grad is False
+                assert self.layer_2[5].weight.requires_grad is True
+                optimizer.step(closure=closure)
```
Comment (carmocca) [inline tests/core/test_lightning_module.py:204]:
```suggestion
            else:
                assert self.layer_1[0].weight.requires_grad is False
                assert self.layer_1[2].weight.requires_grad is False
                assert self.layer_1[4].weight.requires_grad is False

                assert self.layer_2[1].weight.requires_grad is False
                assert self.layer_2[3].weight.requires_grad is False
                assert self.layer_2[5].weight.requires_grad is True
            optimizer.step(closure=closure)
```

File: pytorch_lightning/core/lightning.py [hunk 8]
```diff
@@ -1166,16 +1166,47 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
...
-            param.requires_grad = False
+        param_requires_grad_state = {}
+        # make sure current optimizer is latest to be iterated over.
+        optimizers = [opt for opt in self.optimizers(use_pl_optimizer=False) if opt != optimizer] + [optimizer]
+        num_optimizers = len(optimizers) - 1
+        for opt_idx, opt in enumerate(optimizers):
+            for group in opt.param_groups:
+                for param in group['params']:
+                    if num_optimizers == opt_idx:
+                        # If a param appears in 2 optimizer, revert `requires_grad_state` to before toggle
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:1183]: I think you meant this?

```suggestion
                        # If a param appears in 2 optimizers, revert `requires_grad` to before toggle
```

But how does "appearing in 2 optimizers" relates to this?

Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1183]: We have several tests with 2 optimizers where param are shared. Assuming we are currently running on optimizer_idx = 1, optimizer_idx 0 will save the param state (let's say True) and set its `requires_grad` to `False`.

However, optimizer_idx 1 was expected to see param with `requires_grad=True`.
Therefore, at last, we need to iterate back on current optimizer and if param is in state, apply it.

File: pytorch_lightning/core/lightning.py [hunk 9]
```diff
@@ -1169,16 +1169,47 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
...
-        for param in self.parameters():
-            param.requires_grad = False
+        param_requires_grad_state = {}
+        # make sure current optimizer is latest to be iterated over.
+        optimizers = [opt for opt in self.optimizers(use_pl_optimizer=False) if opt != optimizer] + [optimizer]
+        num_optimizers = len(optimizers) - 1
+        for opt_idx, opt in enumerate(optimizers):
+            for group in opt.param_groups:
+                for param in group['params']:
+                    if num_optimizers == opt_idx:
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1185]: can we include some early break, such as you do not need to continue in the loop after the assignment?

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/core/lightning.py:1169]: Do you plan to do it in this PR?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/core/lightning.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1185]: I don't understand what you mean :)

**Review by justusschock — APPROVED** [review]

File: CHANGELOG.md [hunk 10]
```diff
@@ -29,11 +29,20 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
 - Fixed logging `on_train_batch_end` in a callback with multiple optimizers ([#5521](https://github.com/PyTorchLightning/pytorch-lightning/pull/5521))
+
+
 - Fixed `reinit_scheduler_properties` with correct optimizer ([#5519](https://github.com/PyTorchLightning/pytorch-lightning/pull/5519))
+
+
 - Fixed `val_check_interval` with `fast_dev_run` ([#5540](https://github.com/PyTorchLightning/pytorch-lightning/pull/5540))
 
 
+- Fixed `toggle_optimizer` to reset `requieres_grad` state  ([#5574](https://github.com/PyTorchLightning/pytorch-lightning/pull/5574))
```
Comment (SeanNaren) [inline CHANGELOG.md:43]: There is some weirdness going on in the changelog, this should be put within the unreleased section, and the spaces should be removed (keep original intact)

File: pytorch_lightning/core/lightning.py (hunk 9 above)
Comment (SeanNaren) [reply pytorch_lightning/core/lightning.py:1185]: Don't think it's possible to break out the loop early, this function is ensuring that if `require_grads=False`, that we keep the original state of the parameter before setting `require_grads` False to revert the change.

@tchaton let's say I want to turn off my two optimizers for the same parameters
```python
toggle_optimizer(adam_optimizer, ...)
toggle_optimizer(sgd_optimizer, ...)
```
Will this work? I think as it currently stands the function will leave the shared parameters `require_grad=True` right?

File: CHANGELOG.md (hunk 10 above)
Comment (tchaton) [reply CHANGELOG.md]: Done !

File: pytorch_lightning/core/lightning.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:1185]: It should be

```
toggle_optimizer(adam_optimizer, ...)
untoggle_optimizer(adam_optimizer, ...)
toggle_optimizer(sgd_optimizer, ...)
```

**Review by SeanNaren — APPROVED** [review]

**Review by williamFalcon — APPROVED** [review]

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/core/lightning.py:1170]: @tchaton ^^

File: pytorch_lightning/core/lightning.py [hunk 11]
```diff
@@ -1170,16 +1170,47 @@ def toggle_optimizer(self, optimizer: Optimizer, optimizer_idx: int):
...
+                        # If a param appears in 2 optimizers, revert `requires_grad` to before toggle.
+                        if param in param_requires_grad_state:
+                            param.requires_grad = param_requires_grad_state[param]
```
Comment (ananthsub) [inline pytorch_lightning/core/lightning.py:1189]: @tchaton @awaelchli  the last optimizer's parameters are only looked up from the dictionary, but we haven't added them to the dictionary before this point unless they were contained in another optimizer's params.

Comment (ananthsub) [reply pytorch_lightning/core/lightning.py:1189]: I think it should be something like this instead, which matches what @jain-anshul suggested in #5292 
```
        param_requires_grad_state = {}
        for opt_idx, opt in enumerate(self.optimizers(use_pl_optimizer=False)):
            for group in opt.param_groups:
                for param in group['params']:
                    # If a param appears in multiple optimizers, revert `requires_grad` to before toggle
                    if param in param_requires_grad_state:
                        param.requires_grad = param_requires_grad_state[param]
                    else:
                        param_requires_grad_state[param] = param.requires_grad
                        param.requires_grad = False
        for group in optimizer.param_groups:
            for param in group['params']:
                param.requires_grad = param_requires_grad_state[param]
        self._param_requires_grad_state = param_requires_grad_state
 ```

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (ananthsub) [reply pytorch_lightning/core/lightning.py:1170]: @rohitgr7 @carmocca @tchaton please see #5775

**Comment by phammanhhiep** [comment]
I got similar bug when having 2 optimizers (to train a GAN model), and using option "accumulate_grad_batches" in the trainer (https://pytorch-lightning.readthedocs.io/en/1.3.6/common/trainer.html#trainer-class-api). When I set accumulate_grad_batches = 1, there is no longer error.  Somehow when accumulate_grad_batches > 1, the property requires_grad of the discrimination loss (corresponds to the second optimizer defined in configure_optimizers of LightningModule) is set as False.
